pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/QualifiableNode.java

```diff
@@ -0,0 +1,173 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.ast;
+
+import java.util.Arrays;
+
+/**
+ * Nodes that can be described with a qualified name.
+ *
+ * @author Clément Fournier
+ */
+public interface QualifiableNode {
+
+    /**
+     * Returns a qualified name for this node.
+     *
+     * @return A qualified name.
+     */
+    QualifiedName getQualifiedName();
+
+    /**
+     * Represents Qualified Names for use within PackageStats
+     * TODO make unit tests once the visitor is working to ensure new implementations won't break it
+     */
+    class QualifiedName {
```

Comment:
I'd actually prefer to have this class `QualifiedName` in its own file - it just easier to find then.